tests/Unit/Core/Cart/Calculation/Carrier/RuleWhenChangingCarrierTest.php

```diff
@@ -0,0 +1,174 @@
+                ],
+            ],
+        ],
+        3 => [
+            'name'   => 'carrier 3',
+            'isFree' => false,
+            'ranges' => [
+                1 => [
+                    'from'           => 0,
+                    'to'             => 10000,
+                    'shippingPrices' => [
+                        1 => 5.7, // zoneId => price
+                        2 => 6.2, // zoneId => price
+                    ],
+                ],
+            ],
+        ],
+    ];
+
+    public function testCarrierSpecificCartRuleCorresponding()
+    {
+        // specific setUp
+        $this->setCartCarrierFromFixtureId(2);
+        $this->setCartAddress(1);
+        $this->addProductToCart(1, 1);
+
+        // tests
+        $cartRule = $this->getCartRuleFromFixtureId(2);
+        $result   = $cartRule->checkValidity(\Context::getContext(), false, false);
+        $this->assertFalse($result);
+
+        $cartRule = $this->getCartRuleFromFixtureId(1);
+        $result   = $cartRule->checkValidity(\Context::getContext(), false, false);
+        $this->assertTrue($result);
+        $this->cartRulesInCart[] = $cartRule;
+        $result                  = $this->cart->addCartRule($cartRule->id);
+        $this->assertTrue($result);
+        $cartRules = $this->cart->getCartRules();
+        $this->assertCount(1, $cartRules);
+
+        $result = $cartRule->checkValidity(\Context::getContext(), false, false);
+        $this->assertFalse($result);
+        $this->cartRulesInCart[] = $cartRule;
+        $result                  = $this->cart->addCartRule($cartRule->id);
+        $this->assertFalse($result);
+        $cartRules = $this->cart->getCartRules();
+        $this->assertCount(1, $cartRules);
+    }
+
+    public function testCarrierSpecificCartRuleNotCorresponding()
+    {
+        // specific setUp
+        $this->setCartCarrierFromFixtureId(2);
+        $this->setCartAddress(1);
+        $this->addProductToCart(1, 1);
+
+        // tests
+        $cartRule = $this->getCartRuleFromFixtureId(2);
+        $result   = $cartRule->checkValidity(\Context::getContext(), false, false);
+        $this->assertFalse($result);
+    }
+
+    public function testCarrierSpecificCartRuleRemovedWhenChangingCarrier()
+    {
+        // specific setUp
+        $this->setCartCarrierFromFixtureId(2);
+        $this->setCartAddress(1);
+        $this->addProductToCart(1, 1);
+
+        // tests
+        $cartRule                = $this->getCartRuleFromFixtureId(1);
+        $this->cartRulesInCart[] = $cartRule;
+        $result                  = $this->cart->addCartRule($cartRule->id);
+        $this->assertTrue($result);
+        $cartRules = $this->cart->getCartRules();
+        $this->assertCount(1, $cartRules);
+
+        $this->setCartCarrierFromFixtureId(1);
+
+        $cartRules = $this->cart->getCartRules();
+        $this->assertCount(0, $cartRules);
+    }
+
+    public function testCarrierSpecificCartRuleWithoutCodeSetAndUnset()
```

Comment:
I love tests <3

Reply:
ahah the more explicit, the more maintainable !